File: sdk/python/src/covenant/validator.py
from __future__ import annotations

from pathlib import Path
from typing import Any

import jsonschema
import yaml

from covenant.errors import CovenantViolationError


def _load_schema(spec_dir: Path, schema_path: str) -> dict[str, Any]:
    full = (spec_dir / schema_path).resolve()
    return yaml.safe_load(full.read_text(encoding="utf-8"))


def validate_input(data: Any, schema_path: str | None, spec_dir: Path) -> None:
    """Validate input data against the protocols.input.schema JSON Schema.

    Args:
        data: Input value to validate.
        schema_path: Relative path to the JSON Schema file, or None to skip.
        spec_dir: Directory containing the .covenant.yaml (schema path is relative to this).

    Raises:
        CovenantViolationError: INPUT_SCHEMA_MISMATCH if validation fails.
    """
    if not schema_path:
        return
    schema = _load_schema(spec_dir, schema_path)
    errors = list(jsonschema.Draft7Validator(schema).iter_errors(data))
    if errors:
        raise CovenantViolationError(
            code="INPUT_SCHEMA_MISMATCH",
            detail={"errors": [e.message for e in errors]},
        )


def validate_output(data: Any, schema_path: str | None, spec_dir: Path) -> None:
    """Validate output data against the protocols.output.schema JSON Schema.

    Args:
        data: Output value to validate.
        schema_path: Relative path to the JSON Schema file, or None to skip.
        spec_dir: Directory containing the .covenant.yaml (schema path is relative to this).

    Raises:
        CovenantViolationError: OUTPUT_SCHEMA_MISMATCH if validation fails.
    """
    if not schema_path:
        return
    schema = _load_schema(spec_dir, schema_path)
    errors = list(jsonschema.Draft7Validator(schema).iter_errors(data))
    if errors:
        raise CovenantViolationError(
            code="OUTPUT_SCHEMA_MISMATCH",
            detail={"errors": [e.message for e in errors]},
        )

File: sdk/typescript/src/validator.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import Ajv from "ajv";
import addFormats from "ajv-formats";
import { CovenantViolationError } from "./errors.js";

const _schemaCache = new Map<string, ReturnType<Ajv["compile"]>>();
const _ajv = new Ajv({ allErrors: true, strict: false });
addFormats(_ajv);

function getValidator(schemaPath: string): ReturnType<Ajv["compile"]> {
  const cached = _schemaCache.get(schemaPath);
  if (cached) return cached;
  const schema = JSON.parse(readFileSync(schemaPath, "utf-8")) as object;
  const fn = _ajv.compile(schema);
  _schemaCache.set(schemaPath, fn);
  return fn;
}

/**
 * Validate input data against the protocols.input.schema JSON Schema.
 *
 * @param data - Input value to validate
 * @param schemaPath - Relative path to schema file, or undefined to skip
 * @param specDir - Directory containing the .covenant.yaml
 * @throws CovenantViolationError INPUT_SCHEMA_MISMATCH on failure
 */
export function validateInput(
  data: unknown,
  schemaPath: string | undefined,
  specDir: string
): void {
  if (!schemaPath) return;
  const full = resolve(specDir, schemaPath);
  const validate = getValidator(full);
  if (!validate(data) && validate.errors) {
    throw new CovenantViolationError("INPUT_SCHEMA_MISMATCH", {
      errors: validate.errors.map((e) => e.message ?? "invalid"),
    });
  }
}

/**
 * Validate output data against the protocols.output.schema JSON Schema.
 *
 * @param data - Output value to validate
 * @param schemaPath - Relative path to schema file, or undefined to skip
 * @param specDir - Directory containing the .covenant.yaml
 * @throws CovenantViolationError OUTPUT_SCHEMA_MISMATCH on failure
 */
export function validateOutput(
  data: unknown,
  schemaPath: string | undefined,
  specDir: string
): void {
  if (!schemaPath) return;
  const full = resolve(specDir, schemaPath);
  const validate = getValidator(full);
  if (!validate(data) && validate.errors) {
    throw new CovenantViolationError("OUTPUT_SCHEMA_MISMATCH", {
      errors: validate.errors.map((e) => e.message ?? "invalid"),
    });
  }
}
